geospace-lib/src/main/scala/com/socrata/geospace/lib/shapefile/MultiLayerShapefileReader.scala

```diff
@@ -0,0 +1,160 @@
+package com.socrata.geospace.lib.shapefile
+
+import java.io.{IOException, File}
+
+import com.socrata.geospace.lib.Utils._
+import com.socrata.geospace.lib.errors.InvalidShapefileSet
+import org.apache.commons.io.FilenameUtils
+import org.geoscript.feature._
+import org.geoscript.layer.Shapefile
+import org.geoscript.projection._
+import org.geoscript.feature.schemaBuilder._
+import org.geoscript.feature._
+import org.geoscript.layer._
+import org.geotools.factory.Hints
+import org.geoscript.projection.Projection
+import org.geotools.referencing.ReferencingFactoryFinder
+import org.opengis.referencing.crs.CoordinateReferenceSystem
+
+/**
+ * Shape file reader but for the case where we have multiple layers.
+ * Notable is that we can also pass a different project if we so desire.
+ * If not, then default is take to be "EPSG:4326"
+ * @param projectionString
+ * @param forceLatLon
+ */
+case class MultiLayerShapefileReader(projectionString: String, forceLatLon: Boolean) extends ShapeReader {
+
+  def this(forceLatLon: Boolean) = this(ShapeFileConstants.StandardProjection, forceLatLon)
+
+  val projection = getTargetProjection(projectionString, forceLatLon).fold(throw _, x => x )
+
+  /**
+   * Validates a shapefile and extracts its contents
+   * @param directory Directory containing the set of files that make up the shapefile
+   * @return The shapefile shape layer and schema
+   */
+  def read(directory: File): Either[InvalidShapefileSet, Map[String, (Traversable[Feature], Schema)]] = {
+    // validates the shapefile, returns a failure case if
+    validate(directory).right.flatMap(getContents(_))
+  }
+
+  /**
+   * Validates that the shapefile directory contains the expected set of files and nothing else
+   * @param directory Directory containing the set of files that make up the shapefile
+   */
+  def validate(directory: File): Either[InvalidShapefileSet, Map[String, Array[File]]] = {
```

Comment:
Ah, ok, I missed the earlier comment that we were transitioning away from
SingleLayer.  In that case, I'm fine with leaving it as is, but be sure to
leave a comment or TODO.

I was thinking about the actual processing of the features themselves as
one thing that is pretty much the same, except where the projection comes
from, right?

On Sat, Mar 7, 2015 at 4:19 PM, AEscobarCruz <notifications@github.com>
wrote:

> In
> geospace-lib/src/main/scala/com/socrata/geospace/lib/shapefile/MultiLayerShapefileReader.scala
> <https://github.com/socrata/geospace/pull/43#discussion_r26000955>:
>
> > +
> > +  /**
> > +   * Validates a shapefile and extracts its contents
> > +   * @param directory Directory containing the set of files that make up the shapefile
> > +   * @return The shapefile shape layer and schema
> > +   */
> > +  def read(directory: File): Either[InvalidShapefileSet, Map[String, (Traversable[Feature], Schema)]] = {
> > +    // validates the shapefile, returns a failure case if
> > +    validate(directory).right.flatMap(getContents(_))
> > +  }
> > +
> > +  /**
> > +   * Validates that the shapefile directory contains the expected set of files and nothing else
> > +   * @param directory Directory containing the set of files that make up the shapefile
> > +   */
> > +  def validate(directory: File): Either[InvalidShapefileSet, Map[String, Array[File]]] = {
>
> Sounds good, I may have to spend time to refactor the code though. My
> original thought was to slowly move away from SingleLayer since all it is
> really is a special case of MultiLayer functionality. I wanted, and I guess
> it is being noticed, to make everyone aware of this. Note that multilayer
> also takes in variable projections which could be sent via a parameters in
> a request, so it does a little more and extends our functionality. Is
> spending this time to integrate fully be something everyone is ok with?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/socrata/geospace/pull/43/files#r26000955>.
>



-- 
Evan Chan   |   Principal Engineer, Socrata     |    @evanfchan
*Socrata <http://www.socrata.com/> *| The Data Experience Company™
*Socrata isn't about worldly success, but about changing the world
successfully. This is why I'm excited to be here.*
*What does Socrata do? <https://www.youtube.com/watch?v=yyz86MlkriQ> *(watch
our 2 min video)